File: internal_testsuites/typescript/src/testsuite/store_web_files_test/store_web_files_test.ts
import { ServiceName } from "kurtosis-sdk"
import log from "loglevel";
import { Result, ok, err } from "neverthrow";
import axios from "axios"

import { createEnclave } from "../../test_helpers/enclave_setup";
import {startFileServer} from "../../test_helpers/test_helpers";

const TEST_NAME = "files-artifact-mounting"
const IS_PARTITIONING_ENABLED = false

const FILE_SERVER_SERVICE_NAME: ServiceName = "file-server"

const TEST_FILES_ARTIFACT_URL = "https://kurtosis-public-access.s3.us-east-1.amazonaws.com/test-artifacts/static-fileserver-files.tgz"
const TEST_ARTIFACT_NAME = "test-artifact"


// Filenames & contents for the files stored in the files artifact
const FILE1_FILENAME = "file1.txt"
const FILE2_FILENAME = "file2.txt"

const EXPECTED_FILE1_CONTENTS = "file1\n"
const EXPECTED_FILE2_CONTENTS = "file2\n"

jest.setTimeout(180000)

test.skip("Test web file storing", async () => {
    // ------------------------------------- ENGINE SETUP ----------------------------------------------
    const createEnclaveResult = await createEnclave(TEST_NAME, IS_PARTITIONING_ENABLED)

    if(createEnclaveResult.isErr()) { throw createEnclaveResult.error }

    const { enclaveContext, stopEnclaveFunction } = createEnclaveResult.value

    try {

        // ------------------------------------- TEST SETUP ----------------------------------------------
        const storeWebFilesResult = await enclaveContext.storeWebFiles(TEST_FILES_ARTIFACT_URL, TEST_ARTIFACT_NAME);
        if(storeWebFilesResult.isErr()) { throw storeWebFilesResult.error }
        const filesArtifactUuid = storeWebFilesResult.value;

        const startFileServerResult = await startFileServer(FILE_SERVER_SERVICE_NAME, filesArtifactUuid, FILE1_FILENAME, enclaveContext)
        if(startFileServerResult.isErr()) { throw startFileServerResult.error }
        const {fileServerPublicIp, fileServerPublicPortNum} = startFileServerResult.value
        log.info(`Added file server service with public IP "${fileServerPublicIp}" and port "${fileServerPublicPortNum}"`)

        // ------------------------------------- TEST RUN ----------------------------------------------

        const file1ContentsResult = await getFileContents(
            fileServerPublicIp,
            fileServerPublicPortNum,
            FILE1_FILENAME
        )
        if(file1ContentsResult.isErr()){
            log.error("An error occurred getting file 1's contents")
            throw file1ContentsResult.error
        }

        const file1Contents = file1ContentsResult.value
        if(file1Contents !== EXPECTED_FILE1_CONTENTS){
            throw new Error(`Actual file 1 contents "${file1Contents}" != expected file 1 contents "${EXPECTED_FILE1_CONTENTS}"`)
        }

        const file2ContentsResult = await getFileContents(
            fileServerPublicIp,
            fileServerPublicPortNum,
            FILE2_FILENAME
        )

        if(file2ContentsResult.isErr()){
            log.error("An error occurred getting file 2's contents")
            throw file2ContentsResult.error
        }

        const file2Contents = file2ContentsResult.value
        if(file2Contents !== EXPECTED_FILE2_CONTENTS){
            throw new Error(`Actual file 2 contents "${file2Contents}" != expected file 2 contents "${EXPECTED_FILE2_CONTENTS}"`)
        }
    }finally{
        stopEnclaveFunction()
    }
    jest.clearAllTimers()
})

// ====================================================================================================
//                                       Private helper functions
// ====================================================================================================

async function getFileContents(ipAddress: string, portNum: number, filename: string): Promise<Result<string, Error>> {
    let response;
    try {
        response = await axios(`http://${ipAddress}:${portNum}/${filename}`)
    }catch(error){
        log.error(`An error occurred getting the contents of file "${filename}"`)
        if(error instanceof Error){
            return err(error)
        }else{
            return err(new Error("An error occurred getting the contents of file, but the error wasn't of type Error"))
        }
    }
    const bodyStr = String(response.data)
    return ok(bodyStr)
}

File: internal_testsuites/golang/testsuite/store_web_files_test/store_web_files_test.go
package files_artifact_mounting_test

import (
	"context"
	"fmt"
	"github.com/dzobbe/PoTE-kurtosis-cli/golang_internal_testsuite/test_helpers"
	"github.com/dzobbe/PoTE-kurtosis/api/golang/core/lib/services"
	"github.com/kurtosis-tech/stacktrace"
	"github.com/sirupsen/logrus"
	"github.com/stretchr/testify/require"
	"io"
	"net/http"
	"testing"
)

const (
	testName = "files-artifact-mounting"

	fileServerServiceName services.ServiceName = "file-server"

	testFilesArtifactUrl = "https://kurtosis-public-access.s3.us-east-1.amazonaws.com/test-artifacts/static-fileserver-files.tgz"
	testArtifactName     = "test-artifact-name"

	// Filenames & contents for the files stored in the files artifact
	file1Filename = "file1.txt"
	file2Filename = "file2.txt"

	expectedFile1Contents = "file1\n"
	expectedFile2Contents = "file2\n"
)

// TODO: remove or re-enable this test in future. This functionality seems to be deprecating soon.
func TestStoreWebFiles(t *testing.T) {
	t.Skip()
	ctx := context.Background()

	// ------------------------------------- ENGINE SETUP ----------------------------------------------
	enclaveCtx, _, destroyEnclaveFunc, err := test_helpers.CreateEnclave(t, ctx, testName)
	require.NoError(t, err, "An error occurred creating an enclave")
	defer func() {
		err = destroyEnclaveFunc()
		require.NoError(t, err, "An error occurred destroying the enclave after the test finished")
	}()
	// ------------------------------------- TEST SETUP ----------------------------------------------
	filesArtifactUuid, err := enclaveCtx.StoreWebFiles(context.Background(), testFilesArtifactUrl, testArtifactName)
	require.NoError(t, err, "An error occurred storing the files artifact")
	fileServerPublicIp, fileServerPublicPortNum, err := test_helpers.StartFileServer(ctx, fileServerServiceName, filesArtifactUuid, file1Filename, enclaveCtx)
	require.NoError(t, err, "An error occurred waiting for the file server service to become available")
	logrus.Infof("Added file server service with public IP '%v' and port '%v'", fileServerPublicIp, fileServerPublicPortNum)
	// ------------------------------------- TEST RUN ----------------------------------------------
	file1Contents, err := getFileContents(
		fileServerPublicIp,
		fileServerPublicPortNum,
		file1Filename,
	)
	require.NoError(t, err, "An error occurred getting file 1's contents")
	require.Equal(
		t,
		expectedFile1Contents,
		file1Contents,
		"Actual file 1 contents '%v' != expected file 1 contents '%v'",
		file1Contents,
		expectedFile1Contents,
	)

	file2Contents, err := getFileContents(
		fileServerPublicIp,
		fileServerPublicPortNum,
		file2Filename,
	)
	require.NoError(t, err, "An error occurred getting file 2's contents")
	require.Equal(
		t,
		expectedFile2Contents,
		file2Contents,
		"Actual file 2 contents '%v' != expected file 2 contents '%v'",
		file2Contents,
		expectedFile2Contents,
	)
}

// ====================================================================================================
//
//	Private helper functions
//
// ====================================================================================================
// nolint
func getFileContents(ipAddress string, portNum uint16, filename string) (string, error) {
	resp, err := http.Get(fmt.Sprintf("http://%v:%v/%v", ipAddress, portNum, filename))
	if err != nil {
		return "", stacktrace.Propagate(err, "An error occurred getting the contents of file '%v'", filename)
	}
	body := resp.Body
	defer func() {
		if err := body.Close(); err != nil {
			logrus.Warnf("We tried to close the response body, but doing so threw an error:\n%v", err)
		}
	}()

	bodyBytes, err := io.ReadAll(body)
	if err != nil {
		return "", stacktrace.Propagate(err, "An error occurred reading the response body when getting the contents of file '%v'", filename)
	}

	bodyStr := string(bodyBytes)
	return bodyStr, nil
}
